### PR TITLE
Add XML export and import API functions

### PR DESCRIPTION
**Pull request by gwprice**
These will be used to support export/import of courses containing video
modules that rely on VAL being populated. This does not allow updating
a video (including adding encodings) via import.

JIRA: MA-110

I created a new branch (and, therefore, PR) to avoid clobbering Nimisha's work.

@nasthagiri @BenjiLee Please review

### CODE REVIEW

File: edxval/api.py [hunk 1]
```diff
@@ -123,6 +124,23 @@ def create_profile(profile_name):
...
+    except Video.DoesNotExist:
+        error_message = u"Video not found for edx_video_id: {0}".format(edx_video_id)
+        raise ValVideoNotFoundError(error_message)
+    except Exception:
+        error_message = u"Could not get edx_video_id: {0}".format(edx_video_id)
+        logger.exception(error_message)
+        raise ValInternalError(error_message)
+
+
 def get_video_info(edx_video_id, location=None):  # pylint: disable=W0613
```
Comment (nasthagiri) [inline edxval/api.py]: can we remove the unused parameter?

File: edxval/api.py [hunk 2]
```diff
@@ -365,3 +373,99 @@ def copy_course_videos(source_course_id, destination_course_id):
...
+
+def import_from_xml(xml, edx_video_id, course_id=None):
+    """
+    Imports data from a video_asset element about the given edx_video_id.
+
+    If the edx_video_id already exists, then no changes are made. If an unknown
+    profile is referenced by an encoded video, that encoding will be ignored.
+
+    Args:
+        xml: An lxml video_asset_data element containing import data
```
Comment (nasthagiri) [inline edxval/api.py]: s/video_asset_data/video_asset

File: edxval/api.py [hunk 3]
```diff
@@ -365,3 +373,99 @@ def copy_course_videos(source_course_id, destination_course_id):
...
+        )
+        return
+    except ValVideoNotFoundError:
+        pass
+
+    data = {
+        'edx_video_id': edx_video_id,
+        'client_video_id': xml.get('client_video_id'),
+        'duration': xml.get('duration'),
+        'status': 'imported',
```
Comment (nasthagiri) [inline edxval/api.py]: why is the `status` value not kept during the import?  The AssetMetaDataStore is imported with the course and would contain the uploaded video information.

File: edxval/api.py [hunk 4]
```diff
@@ -365,3 +373,99 @@ def copy_course_videos(source_course_id, destination_course_id):
...
+        profile_name = encoded_video_el.get('profile')
+        try:
+            profile = Profile.objects.get(profile_name=profile_name)
+        except Profile.DoesNotExist:
+            logger.info(
+                "Imported edx_video_id '%s' contains unknown profile '%s'.",
+                edx_video_id,
+                profile_name
+            )
+            continue
```
Comment (nasthagiri) [inline edxval/api.py]: why are we not importing profiles automatically?
by not doing so, it requires manual intervention and careful manual process to make sure profile information is updated before import, etc.  we can be easily flexible about this by automatically creating the profile (and logging as such).

File: edxval/api.py [hunk 5]
```diff
@@ -365,3 +373,99 @@ def copy_course_videos(source_course_id, destination_course_id):
...
+            (optional)
+
+    Raises:
+        ValCannotCreateError: if there is an error importing the video
+    """
+    if xml.tag != 'video_asset':
+        raise ValCannotCreateError('Invalid XML')
+
+    try:
+        video = _get_video(edx_video_id)
```
Comment (nasthagiri) [inline edxval/api.py]: If we are just checking for the pre-existence of the video here, then why call the `_get_video` method that prefetches all the related tables?  It's unnecessarily wieldy for an existence check.

File: edxval/serializers.py [hunk 6]
```diff
@@ -79,8 +79,9 @@ def to_native(self, value):
 
     def from_native(self, data):
         if data:
-            return CourseVideo(course_id=data)
-
+            course_video = CourseVideo(course_id=data)
+            course_video.full_clean(exclude=["video"])
```
Comment (nasthagiri) [inline edxval/serializers.py]: Why was this needed?

File: edxval/tests/test_api.py [hunk 7]
```diff
@@ -687,3 +678,240 @@ def test_existing_video_in_destination_course_id(self):
...
+    def assert_xml_equal(self, left, right):
+        """
+        Assert that the given XML fragments have the same attributes, text, and
+        (recursively) children
+        """
+        def get_child_tags(elem):
+            """Extract the list of tag names for children of elem"""
+            return [child.tag for child in elem]
+
+        for attr in ['tag', 'attrib', 'text', 'tail']:
```
Comment (nasthagiri) [inline edxval/tests/test_api.py]: does this verify the other XML attributes: `'profile', 'url', 'file_size', 'nitrate'`?

File: edxval/tests/test_api.py [hunk 8]
```diff
@@ -687,3 +678,240 @@ def test_existing_video_in_destination_course_id(self):
...
+            **constants.ENCODED_VIDEO_DICT_MOBILE
+        )
+        CourseVideo.objects.create(video=video, course_id='existing_course_id')
+
+    def make_import_xml(self, video_dict, encoded_video_dicts=None):
+        ret = etree.Element(
+            "video_asset",
+            attrib={
+                key: unicode(video_dict[key])
+                for key in ["client_video_id", "duration"]
```
Comment (nasthagiri) [inline edxval/tests/test_api.py]: what about `status`?

File: edxval/api.py (hunk 1 above)
Comment (gwprice) [inline edxval/api.py]: Will do

File: edxval/api.py (hunk 2 above)
Comment (gwprice) [inline edxval/api.py]: Good catch. Will fix.

File: edxval/api.py (hunk 5 above)
Comment (gwprice) [inline edxval/api.py]: Good point. I had used this function for the change with additive merging because the encoded videos and courses were needed, but that is no longer the case.

File: edxval/api.py (hunk 3 above)
Comment (gwprice) [inline edxval/api.py]: The status seems particular to a specific deployment. If something is in progress on edx.org, it is not necessarily in progress on another platform. It seems to me like this is a clear way of indicating that the video resulted from an import, and there will be no further updates without something in the target system specifically acting on the video.

File: edxval/api.py (hunk 4 above)
Comment (gwprice) [inline edxval/api.py]: IIRC, after discussion with the media team, we wanted to leave control over the profiles with them and not with course teams importing data. However, this may be an artifact from back when we didn't want to automatically create profiles with silly default values (for width, height, and extension). I will circle back with them for further discussion.

File: edxval/serializers.py (hunk 6 above)
Comment (gwprice) [inline edxval/serializers.py]: Without this, the course_id is never actually validated. Specifically, ImportTest.test_invalid_course_id fails without this.

File: edxval/tests/test_api.py (hunk 7 above)
Comment (gwprice) [inline edxval/tests/test_api.py]: `attr` here refers to a Python object attribute. The XML attributes are in `attrib` and are compared as a whole dict.

File: edxval/api.py [hunk 9]
```diff
@@ -365,3 +373,99 @@ def copy_course_videos(source_course_id, destination_course_id):
...
+        logger.info(
+            "edx_video_id '%s' present in course '%s' not imported because it exists in VAL.",
+            edx_video_id,
+            course_id,
```
Comment (BenjiLee) [inline edxval/api.py]: Should .format() be used instead?

File: edxval/tests/test_api.py (hunk 7 above)
Comment (nasthagiri) [inline edxval/tests/test_api.py]: ah.. thanks.

File: edxval/serializers.py (hunk 6 above)
Comment (nasthagiri) [inline edxval/serializers.py]: good catch.

File: edxval/api.py (hunk 9 above)
Comment (gwprice) [inline edxval/api.py]: A quick survey of our code reveals that we use both for logging, with the majority using this style. I don't have a strong opinion either way.

File: edxval/tests/test_api.py [hunk 10]
```diff
@@ -687,3 +678,240 @@ def test_existing_video_in_destination_course_id(self):
...
+        """)
+        self.assert_xml_equal(
+            api.export_to_xml(constants.VIDEO_DICT_FISH["edx_video_id"]),
+            expected
+        )
+
+    def test_unknown_video(self):
+        with self.assertRaises(ValVideoNotFoundError):
+            api.export_to_xml("unknown_video")
+
```
Comment (BenjiLee) [inline edxval/tests/test_api.py]: newline. Dunno if this travis will fail you for this.

File: edxval/tests/test_api.py [hunk 11]
```diff
@@ -16,6 +17,7 @@
 from edxval.api import (
     SortDirection,
     ValCannotCreateError,
+    ValVideoNotFoundError,
     VideoSortField,
 )
 from edxval.serializers import VideoSerializer
```
Comment (BenjiLee) [inline edxval/tests/test_api.py]: unused import

File: test-requirements.txt [hunk 12]
```diff
@@ -4,3 +4,4 @@ mock==1.0.1
 django-debug-toolbar==1.2.1
 pylint==1.3.0
 ddt==0.8.0
+lxml==3.3.6
```
Comment (BenjiLee) [inline test-requirements.txt]: Shouldn't this be in `requirements.txt`?

File: edxval/tests/test_api.py (hunk 10 above)
Comment (gwprice) [inline edxval/tests/test_api.py]: Will fix

File: edxval/api.py (hunk 4 above)
Comment (gwprice) [inline edxval/api.py]: I talked with Greg M, and he said that the profiles should be managed by them and not automatically created by course imports.

File: test-requirements.txt (hunk 12 above)
Comment (gwprice) [inline test-requirements.txt]: Yes

File: edxval/tests/test_api.py (hunk 11 above)
Comment (gwprice) [inline edxval/tests/test_api.py]: Good catch

File: edxval/api.py (hunk 3 above)
Comment (gwprice) [inline edxval/api.py]: We discussed this in person and will leave as is if we can make it work reasonably with the video upload page in Studio.

File: edxval/tests/test_api.py [hunk 13]
```diff
@@ -687,3 +677,241 @@ def test_existing_video_in_destination_course_id(self):
...
+
+    def test_unknown_video(self):
+        with self.assertRaises(ValVideoNotFoundError):
+            api.export_to_xml("unknown_video")
+
+
+class ImportTest(TestCase):
+    """Tests import_from_xml"""
+    def setUp(self):
+        mobile_profile=Profile.objects.create(profile_name=constants.PROFILE_MOBILE)
```
Comment (BenjiLee) [inline edxval/tests/test_api.py]: nitpick: `mobile_profile = Profile...`

File: edxval/tests/test_api.py [hunk 14]
```diff
@@ -687,3 +678,240 @@ def test_existing_video_in_destination_course_id(self):
...
+
+        video = Video.objects.get(edx_video_id=constants.VIDEO_DICT_STAR["edx_video_id"])
+        self.assertFalse(video.encoded_videos.filter(profile__profile_name=profile).exists())
+
+    def test_invalid_tag(self):
+        xml = etree.Element(
+            "invalid_tag",
+            attrib={
+                "client_video_id": "dummy",
+                "duration": "0",
```
Comment (BenjiLee) [inline edxval/tests/test_api.py]: Why is this 0 a string?

Comment (gwprice) [inline edxval/tests/test_api.py]: It's an XML attribute

File: edxval/tests/test_api.py (hunk 13 above)
Comment (gwprice) [inline edxval/tests/test_api.py]: Will fix

File: edxval/tests/test_api.py [hunk 15]
```diff
@@ -687,3 +677,241 @@ def test_existing_video_in_destination_course_id(self):
...
+        )
+        self.assert_invalid_import(xml)
+
+    def test_invalid_encoded_video_attr(self):
+        xml = self.make_import_xml(
+            video_dict=constants.VIDEO_DICT_FISH,
+            encoded_video_dicts=[{
+                "url": "http://example.com/dummy",
+                "file_size": -1,
+                "bitrate": 0,
```
Comment (BenjiLee) [inline edxval/tests/test_api.py]: Same test for negative bitrate?

Comment (gwprice) [inline edxval/tests/test_api.py]: I don't think that would provide much additional value. If we test each attribute separately, then we are really just doing an exhaustive test of the VideoSerializer, which should be done elsewhere.

Comment (gwprice) [inline edxval/tests/test_api.py]: OTOH, it would be pretty easy to do with ddt

Comment (BenjiLee) [inline edxval/tests/test_api.py]: Then should this test also be put to said other place. (I'm assuming we are testing the negative filesize)

File: edxval/api.py [hunk 16]
```diff
@@ -365,3 +372,96 @@ def copy_course_videos(source_course_id, destination_course_id):
...
+        'client_video_id': xml.get('client_video_id'),
+        'duration': xml.get('duration'),
+        'status': 'imported',
+        'encoded_videos': [],
+        'courses': [],
+    }
+    for encoded_video_el in xml.iterfind('encoded_video'):
+        profile_name = encoded_video_el.get('profile')
+        try:
+            profile = Profile.objects.get(profile_name=profile_name)
```
Comment (BenjiLee) [inline edxval/api.py]: unused variable `profile`

File: edxval/api.py [hunk 17]
```diff
@@ -365,3 +372,96 @@ def copy_course_videos(source_course_id, destination_course_id):
...
+    video = _get_video(edx_video_id)
+    video_el = Element(
+        'video_asset',
+        attrib={
+            'client_video_id': video.client_video_id,
+            'duration': unicode(video.duration),
+        }
+    )
+    for encoded_video in video.encoded_videos.all():
+         SubElement(
```
Comment (BenjiLee) [inline edxval/api.py]: indent is 5 spaces

File: edxval/tests/test_api.py (hunk 15 above)
Comment (gwprice) [inline edxval/tests/test_api.py]: Those tests already exist

Comment (BenjiLee) [inline edxval/tests/test_api.py]: Ah I see, we are testing a general invalid xml in this case.

File: edxval/api.py (hunk 16 above)
Comment (gwprice) [inline edxval/api.py]: Will fix

File: edxval/api.py (hunk 17 above)
Comment (gwprice) [inline edxval/api.py]: Will fix

**Comment by BenjiLee** [comment]
Happy so far. I'll take another look tonight.

File: edxval/tests/test_api.py [hunk 18]
```diff
@@ -687,3 +677,241 @@ def test_existing_video_in_destination_course_id(self):
...
+            profile=mobile_profile,
+            **constants.ENCODED_VIDEO_DICT_MOBILE
+        )
+        EncodedVideo.objects.create(
+            video=video,
+            profile=desktop_profile,
+            **constants.ENCODED_VIDEO_DICT_DESKTOP
+        )
+
+    def assert_xml_equal(self, left, right):
```
Comment (nasthagiri) [inline edxval/tests/test_api.py]: we should eventually move this to a common utility library.

File: edxval/tests/test_api.py [hunk 19]
```diff
@@ -687,3 +677,241 @@ def test_existing_video_in_destination_course_id(self):
...
+        video = Video.objects.get(edx_video_id=constants.VIDEO_DICT_FISH["edx_video_id"])
+        self.assert_video_matches_dict(video, constants.VIDEO_DICT_FISH)
+        self.assert_encoded_video_matches_dict(
+            video.encoded_videos.get(profile__profile_name=constants.PROFILE_MOBILE),
+            constants.ENCODED_VIDEO_DICT_MOBILE
+        )
+        self.assertFalse(
+            video.encoded_videos.filter(profile__profile_name=constants.PROFILE_DESKTOP).exists()
+        )
+        self.assertFalse(video.courses.filter(course_id=new_course_id).exists())
```
Comment (nasthagiri) [inline edxval/tests/test_api.py]: If the video already existed, for performance reasons, the `course_id` should have still been added to the list of courses.

**Comment by nasthagiri** [comment]
:+1: except for: "If the video already existed, for performance reasons, the course_id should have still been added to the list of courses."

File: edxval/tests/test_api.py (hunk 19 above)
Comment (gwprice) [inline edxval/tests/test_api.py]: That would be nice but is made more complicated by the fact that the CourseSerializer is not in fact a serializer but a field, so I punted on that.

**Comment by BenjiLee** [comment]
obligatory :+1:
